docs/embedded_builds.md

```diff
@@ -0,0 +1,190 @@
+and embed static libraries for all dependencies of bpftrace, rather than
+relying on system libraries.
+
+To build the necessary static LLVM or Clang libraries, `-DEMBED_LLVM:BOOL=ON`
+and `-DEMBED_CLANG=ON` can be used respectively. Clang can link to system LLVM
+as well, but may need to be patched.
+
+It is possible to link bpftrace's embedded libraries with system libraries, so
+linking to a distribution specific LLVM is possible. Many distributions do not
+include `libclang.a`,  using `-DEMBED_LIBCLANG_ONLY=ON` will build only
+`libclang.a`, allowing for linking to system LLVM and Clang libraries on Debian,
+as opposed to Vanilla LLVM sources.
+
+# Build times
+
+To build from scratch on a modern development laptop with an 8-way parallel
+build, bpftrace and its dependencies can be built in about 30 minutes.
+
+Subsequent builds will not have this overhead, once the embedded dependencies
+are built and cached, and will proceed at the comparable speeds as dynamically
+linked builds.
+
+## CI environments
+
+Internal CI environments with any sort of caching mechanism should be able to
+just cache any external project's build directories, and the overhead of
+building on CI shouldn't be any different than other CI jobs.
+
+# Embedding clang
+
+Clang is the actual dependency of bpftrace, as the src/ast folder pulls it in
+for the interface that bpftraces uses for LLVM's code generation platform, which
+ultimately outputs BPF instructions.
+
+On the alpine platform, libclang.a is generated by default and a static build is
+already achieved using this distribution. This libclang dependency is only
+achievable with custom builds, or by using alpine.
+
+To get around this problem, the `embed_clang.cmake` file provides the necessary
+static libraries for bpftrace to link against, by downloading LLVM from a github
+tagged release, and compiling it directly using custom cmake flags, based on
+those provided by alpine already. This basically pulls the fairly simple
+alpine-build "in-tree", using cmake flags based on it, and specifically
+itemizing the libraries that will be link-time dependencies of bpftrace.
+
+In turn, these clang libraries depend heavily on LLVM libraries, so this
+necessitates having access to LLVM static libraries as well.
+
+It is also possible to build only libclang from (mostly) vanilla sources, and
+use system LLVM and Clang libraries. This is what the `EMBED_LIBCLANG_ONLY`
+flag provides.
+
+# Embedding LLVM
+
+It is possible to avoid embedding LLVM by applying distribution specific
+patches to embedded clang, so that it will be patch-compatible with the LLVM
+libraries shipped with the target system.
+
+This can save time by avoiding building LLVM altogether, but as distributions
+LLVM libraries may have more dependencies and other bloat than the embedded
+LLVM, this can result in a larger executable. On Ubuntu, the size of the
+bpftrace executable is increased from 52MB to 88MB, due to the extra libraries
```

Comment:
Maybe add a date so readers have a sense how stale the information is when they read it
```suggestion
bpftrace executable is increased from 52MB to 88MB (measured XX/YY/ZZZZ), due to the extra libraries
```